pallets/ddc-verification/src/lib.rs

```diff
@@ -1757,15 +1851,46 @@ pub mod pallet {
 			valid_aggregates_passed_challenges
 		}
 
+		pub(crate) fn fetch_valid_node_aggregates_passed_challenges(
+			node_aggregates_passed_challenges: Vec<(String, Vec<NodeActivity>)>,
+		) -> Vec<NodeActivity> {
+			let mut valid_aggregates_passed_challenges: Vec<NodeActivity> = vec![];
+
+			for (node_id, node_aggregates_passed_challenge_activities) in
+				node_aggregates_passed_challenges
+			{
+				let valid_activities =
+					node_aggregates_passed_challenge_activities.iter().cloned().max_by_key(
+						|activity| activity.transferred_bytes as i64 + activity.stored_bytes,
+					);
+
+				if let Some(activity) = valid_activities {
+					log::info!(
+						"🚀⛳️ The activity node_id:{:?} with maximum usage, which has passed the challenge. The activity detail is {:?}",
+						node_id,
+						activity
+					);
+					valid_aggregates_passed_challenges.push(activity);
+				}
+			}
+			valid_aggregates_passed_challenges
+		}
+
 		pub(crate) fn find_resulting_hash_from_leafs_and_paths(
 			challenge_responses: Vec<ChallengeAggregateResponse>,
 			cluster_id: &ClusterId,
 			era_id: DdcEra,
-			bucket_id: BucketId,
+			bucket_id: Option<BucketId>,
 			node_id: String,
 		) -> Result<ActivityHash, Vec<OCWError>> {
 			let mut resulting_hash_from_leafs_and_paths = ActivityHash::default();
 
+			if let Some(b_id) = bucket_id {
+				log::info!("🚀find_resulting_hash_from_leafs_and_paths for bucket_id:{:?}", b_id);
+			} else {
```

Comment:
You didn't address my [previous comment](https://github.com/Cerebellum-Network/blockchain-node/pull/434#discussion_r1778198294) on the hash payload. 
Leaves are being hashed differently for `BucketSubTreeLeaf` and `NodeTreeLeaf` structs at the DDC side. 
Here you have only 1 common type of `Leaf` that will not work for both cases
```
	impl Activity for Leaf {
...
                // this will be different depending on Aggregate type
		fn hash<T: Config>(&self) -> ActivityHash {
			let mut data = self.record.id.encode();
			data.extend_from_slice(&self.record.upstream.request.requestType.encode());
			data.extend_from_slice(&self.stored_bytes.encode());
			data.extend_from_slice(&self.transferred_bytes.encode());
			T::ActivityHasher::hash(&data).into()
		}
...
	}
```

Reply:
You are working on processing Node Aggregates in this PR. My [comment](https://github.com/Cerebellum-Network/blockchain-node/pull/434#discussion_r1778198294) was related to both Node Aggregate and Bucket Aggregate (that you worked in previously merged PR) and intended to bring clarity to hashing algorithms used in both kinds of Merkle Trees Nodes and Buckets. You were supposed to implement the `NodeTreeLeaf` in this PR.